src/overlays/actors/ovl_En_Tg/z_en_tg.c

```diff
@@ -76,37 +91,191 @@ static DamageTable D_809901F8 = {
+
+    gDPPipeSync(POLY_OPA_DISP++);
+    gSPSegment(POLY_OPA_DISP++, 0x08, Gfx_EnvColor(play->state.gfxCtx, 0, 50, 160, 0));
+    gSPSegment(POLY_OPA_DISP++, 0x09, Gfx_EnvColor(play->state.gfxCtx, 255, 255, 255, 0));
+
+    SkelAnime_DrawFlexOpa(play, this->skelAnime.skeleton, this->skelAnime.jointTable, this->skelAnime.dListCount,
+                          EnTg_OverrideLimbDraw, EnTg_PostLimbDraw, &this->actor);
+
+    CLOSE_DISPS(play->state.gfxCtx);
+}
+
+/**
+ * Spawns a heart at the first effects array index that's not enabled.
+ * Because of the frame counts, only two hearts are ever spawned at a time.
+ */
+void EnTg_SpawnHeart(EnTg* this, EnTgHeartEffect* effects, Vec3f* heartStartPos, s32 len) {
+    Vec3f heartVelocityVec = { 0.0f, 1.5f, 0.0f };
+    Vec3f zeroVec = { 0.0f, 0.0f, 0.0f };
+    s32 i;
+
+    for (i = 0; i < len && effects->isEnabled; i++, effects++) {}
+
+    if (i < len) {
+        effects->isEnabled = true;
+        effects->pos = *heartStartPos;
+        effects->velocity = heartVelocityVec;
+        effects->unusedZeroVec = zeroVec;
+        effects->scale = 0.01f;
+        effects->pos.x += 4.0f * Math_SinS(this->actor.shape.rot.y);
+        effects->pos.z += 4.0f * Math_CosS(this->actor.shape.rot.y);
+        effects->timer = 16;
+    }
+}
+
+/**
+ * The heart path is curvy as it floats up because of the use of Math_SinS and Math_CosS.
+ */
+void EnTg_UpdateHearts(PlayState* play, EnTgHeartEffect* effects, s32 len) {
+    Vec3f zeroVec = { 0.0f, 0.0f, 0.0f };
+    s16 yaw = Camera_GetInputDirYaw(GET_ACTIVE_CAM(play));
+    s32 i;
+
+    for (i = 0; i < len; i++, effects++) {
+        if (effects->isEnabled == true) {
+            if (DECR(effects->timer) == 0) {
+                effects->isEnabled = false;
+            }
+            effects->pos.y += effects->velocity.y;
+            effects->pos.x += 2.0f * Math_SinS(effects->angle);
+            effects->pos.z += 2.0f * Math_CosS(effects->angle);
+
+            Matrix_Push();
+            Matrix_Translate(effects->pos.x, effects->pos.y, effects->pos.z, MTXMODE_NEW);
+            Matrix_RotateYS(yaw, MTXMODE_APPLY);
+            Matrix_MultVec3f(&zeroVec, &effects->pos);
+            Matrix_Pop();
+
+            effects->angle += 0x1770;
+        }
+    }
+}
+
+void EnTg_DrawHeart(PlayState* play, EnTgHeartEffect* effects, s32 len) {
+    s32 i;
+    s32 flag = false;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Tg/func_8098FBD0.s")
+    OPEN_DISPS(play->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Tg/EnTg_Draw.s")
+    POLY_OPA_DISP = func_801660B8(play, POLY_OPA_DISP);
+    POLY_OPA_DISP = func_8012C724(POLY_OPA_DISP);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Tg/func_8098FD50.s")
+    for (i = 0; i < len; i++, effects++) {
+        if (effects->isEnabled == true) {
+            if (!flag) {
+                gSPDisplayList(POLY_OPA_DISP++, gHoneyAndDarlingHeartMaterialDL);
+                flag = true;
```

Comment:
```suggestion
            if (!isMaterialApplied) {
                gSPDisplayList(POLY_OPA_DISP++, gHoneyAndDarlingHeartMaterialDL);
                isMaterialApplied = true;
```
Something like this might be more descriptive